engine/crates/engine-v2/schema/src/input_value/raw.rs

```diff
@@ -0,0 +1,522 @@
+use crate::{
+    EnumValueId, IdRange, InputValueDefinitionId, RawInputKeyValueId, RawInputObjectFieldValueId, RawInputValueId,
+    SchemaInputValueId,
+};
+
+/// Holds input values for the Schema and for an Operation during execution.
+/// It's generic over 'Str', the string representation:
+/// - StringId for the Schema (default input values and directive arguments)
+/// - Box<str> for the Operation
+///
+/// This allow us to share the code for:
+/// - Display: used to print default values in introspection and add input values in subgraph query strings.
+/// - Serialize: used to serialize variables/arguments.
+/// - Deserializer: used by resolvers to deserialize arguments into a specific struct.
+///
+/// Data is stored in flat arrays for hopefully faster serialization and as as bonus smaller InputValue
+/// footprint on x64 architectures: u64 + usize (padding) rather than 3 * usize (Box<[]> + padding).
+#[derive(Clone)]
+pub struct RawInputValues<Str> {
+    /// Inidividual input values and list values
+    pub(super) values: Vec<RawInputValue<Str>>,
+    /// InputObject's fields
+    pub(super) input_fields: Vec<(InputValueDefinitionId, RawInputValue<Str>)>,
+    /// Object's fields (for JSON)
+    pub(super) key_values: Vec<(Str, RawInputValue<Str>)>,
+}
+
+impl<Str> Default for RawInputValues<Str> {
+    fn default() -> Self {
+        Self {
+            // Reserve the first slot for undefined values.
+            // This allows us in the engine to create from scratch a PlanInputValue that is
+            // undefined. Used when requesting a field argument that wasn't provided.
+            values: vec![RawInputValue::Undefined],
+            input_fields: Default::default(),
+            key_values: Default::default(),
+        }
+    }
+}
+
+/// Represents any possible input value for field arguments, operation variables and directive
+/// arguments.
+/// Used for storage and as input format for operation input values to support default values,
+/// variables, etc.
+#[derive(Debug, Clone)]
+pub enum RawInputValue<Str> {
+    Null,
+    String(Str),
+    EnumValue(EnumValueId),
+    Int(i32),
+    BigInt(i64),
+    Float(f64),
+    Boolean(bool),
+    InputObject(IdRange<RawInputObjectFieldValueId<Str>>),
+    List(IdRange<RawInputValueId<Str>>),
+
+    /// for JSON
+    Map(IdRange<RawInputKeyValueId<Str>>),
+    U64(u64),
+
+    /// Directive arguments have unknown enum values. Likely doesn't make sense, but not sure.
+    UnknownEnumValue(Str),
```

Comment:
This comment could do with an elaboration.  Why do directive arguments have unknown enum values?  What does it mean to be unknown.  In what way doesn't it make sense?

(I do think I know the answer to most of these questions, but others might not)

Reply:
yeah well. Coming from `FederatedGraph` which aren't associated with anything in the Schema, will update rustdoc